File: Sources/StereoSyncCore/Sync/ClockSync.swift
import Foundation

/// Estimates speaker clock offset relative to host using NTP-style exchange.
/// `offset = speakerTime - hostTime`  ⇒  `hostTime ≈ speakerTime - offset`
public struct ClockOffsetEstimate: Sendable {
    public var offset: TimeInterval
    public var roundTrip: TimeInterval
    public var updatedAt: TimeInterval

    public init(offset: TimeInterval, roundTrip: TimeInterval, updatedAt: TimeInterval) {
        self.offset = offset
        self.roundTrip = roundTrip
        self.updatedAt = updatedAt
    }

    /// Convert a host timeline instant into speaker local time.
    public func speakerTime(forHostTime hostTime: TimeInterval) -> TimeInterval {
        hostTime + offset
    }
}

public final class ClockSynchronizer: @unchecked Sendable {
    private let lock = NSLock()
    private var samples: [ClockOffsetEstimate] = []
    private let maxSamples: Int

    public init(maxSamples: Int = 8) {
        self.maxSamples = maxSamples
    }

    public var bestEstimate: ClockOffsetEstimate? {
        lock.lock()
        defer { lock.unlock() }
        // Prefer lowest RTT sample (classic NTP heuristic).
        return samples.min(by: { $0.roundTrip < $1.roundTrip })
    }

    public func recordPong(_ pong: ClockPong, hostReceiveTime: TimeInterval) {
        let rtt = hostReceiveTime - pong.hostSendTime
        guard rtt >= 0, rtt < 2.0 else { return }
        // Speaker midpoint of receive/send approximates remote processing time.
        let speakerMid = (pong.speakerReceiveTime + pong.speakerSendTime) / 2
        let hostMid = (pong.hostSendTime + hostReceiveTime) / 2
        let offset = speakerMid - hostMid
        let estimate = ClockOffsetEstimate(offset: offset, roundTrip: rtt, updatedAt: hostReceiveTime)

        lock.lock()
        samples.append(estimate)
        if samples.count > maxSamples {
            samples.removeFirst(samples.count - maxSamples)
        }
        lock.unlock()
    }

    public func reset() {
        lock.lock()
        samples.removeAll()
        lock.unlock()
    }
}

public enum HostTime {
    /// Monotonic-ish wall clock suitable for demo sync (ProcessInfo uptime).
    public static func now() -> TimeInterval {
        ProcessInfo.processInfo.systemUptime
    }
}

File: android/core/sync/src/main/kotlin/com/chorus/core/sync/ClockSync.kt
package com.chorus.core.sync

import com.chorus.core.protocol.ClockPongData

data class ClockOffsetEstimate(
    val offset: Double,
    val roundTrip: Double,
    val updatedAt: Double,
) {
    fun speakerTimeForHostTime(hostTime: Double): Double = hostTime + offset
}

/**
 * Sliding-window NTP-style offset estimator (host-side samples).
 * Speakers typically apply Host-pushed [ControlPayload.ClockOffset] directly.
 */
class ClockSynchronizer(private val maxSamples: Int = 8) {
    private val lock = Any()
    private val samples = ArrayList<ClockOffsetEstimate>()

    val bestEstimate: ClockOffsetEstimate?
        get() = synchronized(lock) {
            if (samples.isEmpty()) null
            else samples.minBy { it.roundTrip }
        }

    fun recordPong(pong: ClockPongData, hostReceiveTime: Double) {
        val rtt = hostReceiveTime - pong.hostSendTime
        if (rtt < 0 || rtt >= 2.0) return
        val speakerMid = (pong.speakerReceiveTime + pong.speakerSendTime) / 2.0
        val hostMid = (pong.hostSendTime + hostReceiveTime) / 2.0
        val estimate = ClockOffsetEstimate(speakerMid - hostMid, rtt, hostReceiveTime)
        synchronized(lock) {
            samples.add(estimate)
            if (samples.size > maxSamples) {
                samples.subList(0, samples.size - maxSamples).clear()
            }
        }
    }

    fun reset() {
        synchronized(lock) { samples.clear() }
    }
}

/**
 * Monotonic clock in seconds.
 * Prefer [android.os.SystemClock.elapsedRealtimeNanos] on Android via [AndroidHostTime].
 * JVM fallback uses nanoTime for unit tests.
 */
object HostTime {
    @Volatile
    private var provider: () -> Double = { System.nanoTime() / 1_000_000_000.0 }

    fun install(provider: () -> Double) {
        this.provider = provider
    }

    fun now(): Double = provider()
}
